proxy/grpcproxy/chan_stream.go

```diff
@@ -116,17 +116,24 @@ func (s *chanStream) SendMsg(m interface{}) error {
 
 func (s *chanStream) RecvMsg(m interface{}) error {
 	v := m.(*interface{})
-	select {
-	case msg, ok := <-s.recvc:
-		if !ok {
-			return grpc.ErrClientConnClosing
+	for {
+		select {
+		case msg, ok := <-s.recvc:
+			if !ok {
+				s.cancel()
```

Comment:
That bit is just extra paranoia. The problem from recvc being buffered and doing:
```go
ch <- x
cancel()
```
which leads to the `select{}` here having both `<-s.recvc` and `<-s.ctx.Done()` available for reading. Since go won't prioritize selects, it would randomly choose `s.ctx.Done` over `s.recvc` and return a context canceled error.